core/chaincode/lifecycle/deployedcc_infoprovider.go

```diff
@@ -103,6 +103,39 @@ func (vc *ValidatorCommitter) ChaincodeInfo(channelName, chaincodeName string, q
 	}, nil
 }
 
+// AllChaincodesInfo returns a list of deployed chaincodes info, including both new lifecycle and legacy chaincodes
+func (vc *ValidatorCommitter) AllChaincodesInfo(channelName string, sqe ledger.SimpleQueryExecutor) ([]*ledger.DeployedChaincodeInfo, error) {
```

Comment:
Thanks Wenjian, could you please explain the difference between this function and 

https://github.com/hyperledger/fabric/blob/master/core/chaincode/lifecycle/cache.go#L64

It too delivers a comprehensive list of all chaincode definitions from both legacy and new lifecycle.

Reply:
**Background:** When CouchDB is used as the state database, for each chaincode within a channel and for each collection within a chaincode, we create a separate database.

**Scenario:** We want to delete all databases associated with a channel or export data of a particular channel for a ledger snapshot -- 

**Problem:** We don't explicitly store all databases name associated with each channel. Even by querying all databases present in a CouchDB, we cannot identify that. 

**Challenge:** In order to find all existing chaincodes and collections belong to a channel, ledger needs to depend on the lifecycle pkg. 

**Solution:** For all existing channels/chaincodes/collections as well as for new ones, we plan to store the following mapping in CouchDB: [database name, channelName], [database name, chaincodeName] [database name,  collectionName]. 

In  #1268, we started to store the mapping for new databases (also for existing databases but only when they are opened -- quite complicated to explain -- refer to this [discussion](https://github.com/hyperledger/fabric/pull/1268#pullrequestreview-413057258)). 

To store the mapping for all existing databases, currently, we plan to use this API introduced by this PR when the CouchDB database is opened for an existing channel. In other words, when the peer starts up, it initializes various components. As you know, Peer also acquires `ledger.PeerLedger` for each existing channel. When the ledger associated with a channel is opened, we open respective databases too. After opening the database (approximately [here](https://github.com/hyperledger/fabric/blob/master/core/ledger/kvledger/txmgmt/privacyenabledstate/db.go#L98), we want to call the deployedCCInfo provider to get all deployed chaincode names and respective collection names. Once it is retrieved, the db would be asked to update the mapping it maintains (introduced in #1268). While this occurs, the lifecycle cache would not have been populated. This is because we initialize the lifecycle cache at https://github.com/hyperledger/fabric/blob/master/core/ledger/kvledger/kv_ledger.go#L166 once the database is opened for the ledger. 

Further, the `ledger.DeployedCCInfoProvider` is already passed to the kvledger initializer. We plan to reuse the existing dependency itself for our needs. It may not be good to add more dependency to the kvledger package. Note that this API is used once when the peer is upgraded to 2.2 or from old LTS to new LTS. Once we store the mapping of all existing channels/chaincodes/collection to database name, this API can be removed.

Reply:
> https://github.com/hyperledger/fabric/blob/master/core/chaincode/lifecycle/cache.go#L64
> 
> It too delivers a comprehensive list of all chaincode definitions from both legacy and new lifecycle.

As far as I recall, this cache works only maintains data on `_lifecycle` namespace. 

Moreover, the purpose of this cache was primarily to maintain some states so as to be able to capture and process the events related to lifecycle operations, as opposed to what the deployedccinfoprovider (in the same package) does - i.e., provide a stateless set of functions just to hide the details of the lifecycle data model from other components.

Reply:
The lifecycleCache is being used by few other components too (to get endorsement info, I think).

Reply:
thanks @cendhu for the very thorough explanation! But to @manish-sethi 's point: I'm pretty sure that
https://github.com/hyperledger/fabric/blob/master/core/chaincode/lifecycle/cache.go#L64
gives a list of {legacy U new} definitions, with shadowing rules (e.g. if a chaincode is defined in both, we'll get defintions from the new only). Couldn't we just make use of that and register a new listener there? My point is just that we already have a ton of these lifecycle info providers, and we should try to avoid adding new ones unless it's strictly mandatory.

Reply:
@ale-linux thanks for your comment and pointing out cache.go. I see a couple of different usages between  ledger.DeployedChaincodeInfoProvider and lifecycle.Cache.
- DeployedChaincodeInfoProvider is an existing interface used by ledger. This PR adds a new method that returns an existing data type. It is a simple extension and fits the existing usage pattern. On the other hand, ledger does not depend on lifecycle.Cache to retrieve chaincode info, so we will have to introduce new dependency to ledger.
- ledger needs to get chaincode data for both new lifecyle and legacy chaincodes. Looking at https://github.com/hyperledger/fabric/blob/master/internal/peer/node/start.go#L769, it seems that lifecycle cache only inits new lifecycle chaincodes.
- ledger needs to get chaincode data during ledger initialization. The lifecycle cache is initialized after ledger init. Although we may be able to refactor the code, it will add more complexity. 

Given the above, I think it makes sense to add a simple method to DeployedChaincodeInfoProvider instead of introducing lifecycle cache dependency to ledger.

Reply:
Thanks @wenjianqiao !